test/master-playlist-controller.test.js

```diff
@@ -368,6 +368,61 @@ QUnit.test('updates the enabled track when switching audio groups', function(ass
            'enabled a track in the new audio group');
 });
 
+QUnit.test('detects if the player is stuck at the playlist end', function(assert) {
+  let playlistCopy = Hls.Playlist.playlistEnd;
+
+  this.masterPlaylistController.mediaSource.trigger('sourceopen');
+  this.standardXHRResponse(this.requests.shift());
+  let playlist = this.player.tech_.hls.selectPlaylist();
+
+  // not stuck at playlist end when no seekable, even if empty buffer
+  // and positive currentTime
+  this.masterPlaylistController.seekable = () => videojs.createTimeRange();
+  this.player.tech_.buffered = () => videojs.createTimeRange();
+  this.player.tech_.setCurrentTime(170);
+  assert.ok(!this.masterPlaylistController.stuckAtPlaylistEnd_(playlist), 'not stuck at playlist end');
+
+  // not stuck at playlist end when no seekable, even if empty buffer
+  // and currentTime 0
+  this.player.tech_.setCurrentTime(0);
+  assert.ok(!this.masterPlaylistController.stuckAtPlaylistEnd_(playlist), 'not stuck at playlist end');
+
+  // not stuck at playlist end when no seekable but current time is at
+  // the end of the buffered range
+  this.player.tech_.buffered = () => videojs.createTimeRange(0, 170);
+  assert.ok(!this.masterPlaylistController.stuckAtPlaylistEnd_(playlist), 'not stuck at playlist end');
+
+  // not stuck at playlist end when currentTime not at seekable end
+  // even if the buffer is empty
+  this.masterPlaylistController.seekable = () => videojs.createTimeRange(0, 130);
+  this.player.tech_.setCurrentTime(50);
+  this.player.tech_.buffered = () => videojs.createTimeRange();
+  Hls.Playlist.playlistEnd = () => 130;
+  assert.ok(!this.masterPlaylistController.stuckAtPlaylistEnd_(playlist), 'not stuck at playlist end');
```

Comment:
Since `stuckAtPlaylistEnd_` only uses seekable to check if we have a seekable range, but doesn't use `seekable.end(0)` for any of its checks, is there even a point to checking `not stuck at playlist end when currentTime not at seekable end even if the buffer is empty`?

If we still want to keep this part in, I'm not sure having `playlistEnd` and `seekable.end(0)` both be the same (130) is representative of an actual scenario since `seekable.end` is the time since playback started to minus 3 segment durations from the end of the window, where `playlistEnd` is time since playback started to minus 0 segment durations from the end of the window.